src/main/java/redis/clients/jedis/args/Endpoint.java

```diff
@@ -0,0 +1,69 @@
+package redis.clients.jedis.args;
+
+import redis.clients.jedis.Protocol;
+
+public interface Endpoint<T> extends Rawable {
+
+  Endpoint<T> inclusive();
+
+  Endpoint<T> exclusive();
+
+  public static abstract class AbstractEndpoint<T> implements Endpoint<T> {
+
+    private final byte[] raw;
+    private boolean exclusive;
+
+    public AbstractEndpoint(T data, boolean exclusive) {
+      if (data instanceof Rawable) {
+        this.raw = ((Rawable) data).getRaw();
+      } else if (data instanceof Integer) {
+        this.raw = Protocol.toByteArray((Integer) data);
+      } else if (data instanceof Double) {
+        this.raw = Protocol.toByteArray((Double) data);
+      } else {
+        throw new IllegalArgumentException(); // TODO message
+      }
+      this.exclusive = exclusive;
+    }
+
+    public AbstractEndpoint(T data) {
+      this(data, false);
+    }
+
+    @Override
+    public Endpoint<T> exclusive() {
+      exclusive = true;
+      return this;
+    }
+
+    @Override
+    public Endpoint<T> inclusive() {
+      exclusive = false;
+      return this;
+    }
+
+    @Override
+    public byte[] getRaw() {
+      if (!exclusive) {
+        return raw;
+      }
+      byte[] b = new byte[1 + raw.length];
+      b[0] = EXCLUSIVE_PREFIX;
+      System.arraycopy(raw, 0, b, 1, raw.length);
+      return b;
+    }
+  }
+
+  static final byte EXCLUSIVE_PREFIX = '(';
+
+  public static Endpoint<Double> of(double value) {
+    return new AbstractEndpoint<Double>(value) {
+    };
+  }
+
+  public static Endpoint<Double> of(double value, boolean exclusive) {
+    return new AbstractEndpoint<Double>(value, exclusive) {
+    };
+  }
+
+}
```

Comment:
1. I think the name `Endpoint` is really confusing, how about `RangePoint` or `RangeLimit` or anything else (I suck at naming things :D)
2. I think the class is overcomplicated a bit here, can we reduce to just what we need? Specifically
   1. I don't see why we need an interface at all
   2. Not really sure we should support `Rawable` as a data type here, seems a bit risky given the scope of `Rawable` (e.g. I can make an `Endpoint<Command>` but that doesn't make much sense)
   2. The Abstract class is not really abstract here
   3. I don't see the need for the mutability using `inclusive()` and `exclusive()` functions
   4. We don't need a public constructor we can just expose static helper constructors

```suggestion
public final class RangePoint<T> implements Rawable {

  public static <T> RangePoint<T> exclusive(T data) {
    return new RangePoint<>(data, true);
  }

  public static <T> RangePoint<T> inclusive(T data) {
    return new RangePoint<>(data, false);
  }

  private final byte[] raw;

  private RangePoint(T data, boolean exclusive) {
    byte[] temp;

    if (data instanceof Integer) {
      temp = Protocol.toByteArray((Integer) data);
    } else if (data instanceof Double) {
      temp = Protocol.toByteArray((Double) data);
    } else {
      throw new IllegalArgumentException("RangePoint does not support data type: " + data.getClass().getName());
    }

    if (exclusive) {
      this.raw = new byte[1 + temp.length];
      this.raw[0] = '(';
      System.arraycopy(temp, 0, this.raw, 1, temp.length);
    } else {
      this.raw = temp;
    }
  }

  @Override
  public byte[] getRaw() {
    return this.raw;
  }
}
```

Reply:
@mina-asham 

About naming: How about `RangeEndpoint`?

About implementation: I used interface so that users can have their own version. I added abstract class to ease that process.

Reply:
@mina-asham 

About naming: How about `RangeEndpoint`?

About implementation: I used interface so that users can have their own version. I added sample (abstract) class and public constructor to ease that process.

About abstract class: True. We can rename it to `EndpointBase`/`EndpointImpl` or something like that.

About mutability: I think it'd be better to have that.

Reply:
> About naming: How about RangeEndpoint?

I am okay with that one :)

> About implementation: I used interface so that users can have their own version. I added sample (abstract) class and public constructor to ease that process.

> About mutability: I think it'd be better to have that.

For both of these, yes in abstract it's nice to have functionality, but why do we really need it? Providing more than needed can the library to get bloated with functionality that no one really needs, making further enhancements updates harder due to having to support some backward comptability issues. If you have a strong use case (e.g. why would I need to mutate a range endpoint into exclusive/inclusive? why would I want to implement my own solution or what is missing from this one?)  to any of that I am all for it but I personally think we shouldn't do it otherwise.

Reply:
@sazzad16 I think @mina-asham has a point here, what use case did you image going with the abstract class?

Reply:
@gkorland Customizability; e.g. for modules (in case something comes up) or user's custom implementations.